[FSDP][optim_state_dict] Consolidate rank0_only load logic

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99731
* __->__ #99647
* #99624

Follow up https://github.com/pytorch/pytorch/pull/99624, this PR consolidate the logic of `use_orig_params=False` with `use_orig_params=True` to use the same logic to load optimizer checkpoint when rank0_only is True.

fegin committed:
[FSDP][optim_state_dict] Consolidate rank0_only load logic

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Consolidate rank0_only load logic"




[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Consolidate rank0_only load logic"




[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Consolidate rank0_only load logic"




[ghstack-poisoned]

wz337 commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -493,6 +490,7 @@ def _flatten_optim_state_dict(
                     else:
                         # Move the tensor in the original osd back to CPU to make the
                         # original osd unaffected.
+                        continue
```
QQ about the continue: Are we just skipping moving the state back to CPU here?

fegin replied:
Oop, sorry, didn't remove this before submitting the PR. Will update it.

wz337 reviewed (approved):
LGTM!

fegin committed:
Update on "[FSDP][optim_state_dict] Consolidate rank0_only load logic"


Follow up https://github.com/pytorch/pytorch/pull/99624, this PR consolidate the logic of `use_orig_params=False` with `use_orig_params=True` to use the same logic to load optimizer checkpoint when rank0_only is True.



[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Consolidate rank0_only load logic"


Follow up https://github.com/pytorch/pytorch/pull/99624, this PR consolidate the logic of `use_orig_params=False` with `use_orig_params=True` to use the same logic to load optimizer checkpoint when rank0_only is True.



[ghstack-poisoned]

fegin commented:
@pytorchbot merge
